arm: Placing the functions which holds __ramfunc into '.ramfunc'

qianfan-Zhao opened the pull request:
__ramfunc is an intrinsis of IAR, using this places a function in RAM
instead of Flash. Code that for example reprograms flash at runtime
can't execute from flash, in that case must placing code into RAM.

This commit create a new section named '.ramfunc' in link scripts,
all functions has __ramfunc keyword saved in thats sections and
will load from flash to sram after the system booted.

Fixes: #10253

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
arm: Placing the functions which holds __ramfunc into '.ramfunc'

Using __ramfunc to places a function in RAM instead of Flash.
Code that for example reprograms flash at runtime can't execute
from flash, in that case must placing code into RAM.

This commit create a new section named '.ramfunc' in link scripts,
all functions has __ramfunc keyword saved in thats sections and
will load from flash to sram after the system booted.

Fixes: #10253

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao commented:
I copied the idea in PR #11199 @aurel32 could you please review this PR?

This PR is based on #10255, I can't reopen it due to had forced push, so I create a new one.
@pfalcon  @SebastianBoe Could you please review this PR again?

CC: @easonxiang @hongshui3000 @walter-xie

SebastianBoe reviewed (dismissed):
Transferring NACK from https://github.com/zephyrproject-rtos/zephyr/pull/10255 to this PR.

qianfan-Zhao commented:
@erwango could you please help me tested this patch on a cortex m7 based board with cache enabled?

SebastianBoe commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,12 @@
+		__ramfunc_ram_end = .;
+	} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
+
+	__ramfunc_rom_start = LOADADDR(.ramfunc);
```
We need a newline here.

SebastianBoe commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,12 @@
+	SECTION_DATA_PROLOGUE(.ramfunc,,)
```
The indentation here should be removed.

SebastianBoe commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -366,6 +373,18 @@ static int arm_mpu_init(struct device *arg)
+	region_conf.base = (u32_t)&__ramfunc_ram_start;
+	if (size > 0) {
+		_region_init(index, &region_conf);
+	}
```
I can't review this.

Would you be able to @ioannisg?

ioannisg commented:
Quick comments  (while still thinking about it)

1. If this is a code region, shouldn't it rather be fixed partition (e.g. like flash0) and have its description in the ARM dts, instead? And use a fixed region, as we do for FLASH regions at init?
2. wouldn't you want to support this for the NXP_MPU, too?
3. wouldn't you want to support thins for ARMv8-M too?

pfalcon commented:
> wouldn't you want to support this for the NXP_MPU, too?
> wouldn't you want to support thins for ARMv8-M too?

And wouldn't we want to think about userspace/kernel split in general and e.g. invite @andrewboie for a consult?

I personally would be ok to go forward with this in "quick and easy" way, but we all should understand that it probably doesn't end there, and will require further elaborate and maintenance.

pfalcon commented:
> __ramfunc is an intrinsis of IAR

And I'd suggest to remove references to propriatary toolchains from the commit message, it's not relevant here. It's a feature in wide use, and "ramfunc" is of course "RAM function", and "__" is a common attribute prefix used in Zephyr.

nashif commented:
> I'd suggest to remove references to propriatary toolchains from the commit message

why?

pfalcon commented:
> why?

I guess "not relevant" argument given above should be enough.

andrewboie commented:
This looks very similar to #11199 

I am thinking we need a more generic way of specifying "MPU regions that are configured to some policy at boot which depend on some boundaries defined by the linker" since that's what both patches are trying to solve.

However I agree with @ioannisg in that I'm curious why we need this to depend on linker boundaries in the first place, rather than use a fixed region, which won't require any additional infrastructure, just another entry in the boot time MPU configuration table.

@aurel32 any opinion?

aurel32 commented:
> I am thinking we need a more generic way of specifying "MPU regions that are configured to some policy at boot which depend on some boundaries defined by the linker" since that's what both patches are trying to solve.

@andrewboie. I agree this the "nocache" and "ramfunc" PR are very similar in the sense they want to reserve some part of the RAM to place some specific object with different MPU permissions. I believe we'll see at some point request for "itcm" and "dtcm" memory for the Cortex-M7, which are also a bit similar although there the address range is not taken from real RAM and fixed. We also already have partial support for "ccm", the linker part is there, but not the MPU part.

All of them looks relatively similar and finding a common solution to handle them is a good idea.

> However I agree with @ioannisg in that I'm curious why we need this to depend on linker boundaries in the first place, rather than use a fixed region

I believe it's not really a problem to use a fixed range for "ramfunc" regions, a given application is available to evaluate how much memory it needs. The "nocache" regions will be used by drivers, it is not really difficult to determine how much "nocache" memory will need driver foo or bar (although that might depend on Kconfig values), but ideally we should be able to configure the region size depending on which driver is enabled.

> which won't require any additional infrastructure, just another entry in the boot time MPU configuration table.

I don't really get that part. I agree we could just add an entry in the boot time MPU configuration table, but we still need a linker entry to place the objects at the correct RAM location.

andrewboie commented:
> I don't really get that part. I agree we could just add an entry in the boot time MPU configuration table, but we still need a linker entry to place the objects at the correct RAM location.

Agreed, you'll have to add stuff to the linker script no matter what, but it would be nice if we could add regions like this without having additional C code in the MPU driver each time.

qianfan-Zhao commented:
@ioannisg 
> If this is a code region, shouldn't it rather be fixed partition (e.g. like flash0) and have its description in the ARM dts, instead? And use a fixed region, as we do for FLASH regions at init?

Seems can't do that. The regions like FLASH_0 and SRAM_0 has the const address and size, so that's can write into a fixed MPU partition. But we can't know which is the address of ramfunc section, and how much size it has.

> wouldn't you want to support this for the NXP_MPU, too?
> wouldn't you want to support thins for ARMv8-M too?

I don't have a board with NXP_MPU or ARMv8-M, I think I can't test that.

ioannisg commented:
> 
> 
> > I don't really get that part. I agree we could just add an entry in the boot time MPU configuration table, but we still need a linker entry to place the objects at the correct RAM location.
> 
> Agreed, you'll have to add stuff to the linker script no matter what, but it would be nice if we could add regions like this without having additional C code in the MPU driver each time.

@andrewboie @aurel32 , my main point was that, here, we have an SRAM region, in which we, apparently, only want to place code, so it is natural that this SRAM is excluded from the image ram (kernel, application and shared mem) completely.

qianfan-Zhao commented:
Minors updates:
@SebastianBoe I had addressed yours reviews.

The ARM MPU request region aligns, that's maybe cause lots of memory wasted on such soc doesn't support MPU. So I had changed the ramfunc.ld, align the ramfunc section only when CONFIG_CPU_HAS_MPU defined.

ioannisg commented:
> The ARM MPU request region aligns, that's maybe cause lots of memory wasted on such soc doesn't support MPU. So I had changed the ramfunc.ld, align the ramfunc section only when CONFIG_CPU_HAS_MPU defined.

I think the right conditional should be the "user" selection of ARM MPU, i.e. CONFIG_ARM_MPU

Still, i think we need to have a generic solution that covers at least the ARM NXP MPU, as well.
If this patch gets accepted, I would very much like to make it cover also the ARMv8-M MPU.

ioannisg reviewed (commented):
Added comments in the discussion.
I think we first need to discuss/review the usefulness of this patch; then we can review the specifics, e.g. linker architecture, Kconfig symbol names, implementation etc.

I am, generally, OK with the principle of having a code RAM section for XIP images.

ioannisg commented:
> 
> 
> @ioannisg
> 
> > I think we first need to discuss/review the usefulness of this patch
> 
> Do you mean usefulness of the feature this patch introduces? IMHO, that's pretty basic feature for deeply embedded systems, with at least to usecases:
> 
>     1. Flash usually have waitstates for high CPU MHZ, while SRAM runs on native CPU speed, so funcs in RAM are useful for performance-critical sections of code.
> 
>     2. As mentioned my @qianfan-Zhao in the commit description, while flash is reprogrammed (page-written, erased), code usually can't run from flash, so real-time ISRs may need to reside in RAM.
> 
> 
> So, the feature is definitely +1 from me. Can't say much about the implementation (like, implementation is trivial in a conventional embedded setup, but with MPUs, kernel/userspace split, memory domains, and stuff, it gets quite involved.)

I think I was clear enough, above, that I also support this patch.
I'd like the other reviewers to comment, too.

aurel32 commented:
I am also in favor of adding a way to support code in RAM for XIP images. Now I agree we have to find a way to do that properly without duplicating nor complexifying the code too much. Add the linker level, I wonder if we can define the section and the start/end pointers by using a macro. This macro can be defined differently if the MPU is enabled or not, or if the MPU requires power of 2 alignment or 32 bytes or any other value.

At the MPU configuration level, things seems a bit more complicated if we want to have variable size sections and support ARMv7-M, ARMv8-M and NXP MPUs.

aurel32 commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -106,6 +106,16 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
+{
+	ARG_UNUSED(base);
+
+	p_attr->rasr = _get_region_attr(0, RO, 0, 1, 0, 0, 0, size);
```
I don't think this is correct. You indeed want to disable the eXecute Never bit, but you want to keep TEX=1, C=1 and B=1.

ioannisg replied:
You should consult the ARMv7-M architecture/technical reference manual.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -106,6 +106,16 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
 }
 
+#ifdef CONFIG_RAM_FUNCTION
+static inline void _get_mpu_ramfunc_region_attr(arm_mpu_region_attr_t *p_attr,
```
Can you implement the function in arm_mpu_v8m_internal.h?

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -106,6 +106,16 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
+
+	p_attr->rasr = _get_region_attr(0, RO, 0, 1, 0, 0, 0, size);
+}
+#endif
```
Add /* xxx */ inline

qianfan-Zhao replied:
means just leave a /* xxx */ comments here?

ioannisg replied:
I mean, /* CONFIG_RAM_FUNCTION */ or whatever will be the name of the Kconfig option, after the review.

ioannisg replied:
And  I meant, on line 118: #endif /* CONFIG_RAM_FUMCTION */
(so it is clear what this #endif refers to)

Sorry for the confusion

ioannisg commented on include/linker/linker-defs.h:
```diff
@@ -249,6 +249,11 @@ extern char _vector_end[];
 /* end address of image, used by newlib for the heap */
 extern char _end[];
 
+/* ramfunc sections */
```
Can you add some explanatory text, for example, as the one for _app_ram_start (lines 184 - 190)?

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -415,6 +415,9 @@ SECTIONS
 
     __data_ram_end = .;
 
+#ifdef CONFIG_RAM_FUNCTION
```
Is there a particular reason why the RAM code section needs to be at the end, i.e. after kernel SRAM?

I would expect it is simpler if you sneak it in the beginning of SRAM.

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,17 @@
+SECTION_DATA_PROLOGUE(.ramfunc,,)
+{
+#ifdef CONFIG_CPU_HAS_MPU
```
As I said, this is not the right Kconfig option.
You could use #if CPU_HAS_ARM_MPU (which is a user option to build with MPU support)

Otherwise, you could just do nothing, and simply align with MPU_GUARD_ALIGN_AND_SIZE in line 10)

ioannisg replied:
Actually, I would prefer the second option, i.e. no #ifdef CPU_HAS_ARM_MPU, but only the alignments, depending on MPU_GUARD_ALIGN_AND_SIZE and MPU_REQUIRE_SPOWER_OF_TWO_ALIGNMENT

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,17 @@
+SECTION_DATA_PROLOGUE(.ramfunc,,)
```
@SebastianBoe could you review just the linker script file?

SebastianBoe replied:
LGTM

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,17 @@
+	KERNEL_INPUT_SECTION(".ramfunc.*")
+#ifdef CONFIG_CPU_HAS_MPU
+	. = ALIGN(32);
+	. = ALIGN( 1 << LOG2CEIL(__ramfunc_ram_end - __ramfunc_ram_start));
```
This one is only required, if the MPU requires power-of-two alignment, IMHO.

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,17 @@
+	. = ALIGN(32);
+	. = ALIGN( 1 << LOG2CEIL(__ramfunc_ram_end - __ramfunc_ram_start));
+#endif
+	__ramfunc_ram_end = .;
```
In order to make sure that the end of the MPU region does not overlap with the adjacent one after that, you need to align the end address as well with the minimum MPU alignment.

ioannisg commented on kernel/Kconfig:
```diff
@@ -195,6 +195,13 @@ config APPLICATION_MEMORY
 	  will have the project-level application objects and any libraries
 	  including the C library in it.
 
+config RAM_FUNCTION
```
- I would rather name this option as XIP_EXECUTE_FROM_RAM, instead, to reflect what we are actually enabling (or something similar)
- @andrewboie  Do we need to have a dependence on XIP? 
- Also, this is not, really, a Kernel option - rather a generic arch option, right? If you agree, I suggest to move the option to arch/Kconfig

qianfan-Zhao replied:
Current __ramfunc only support ARM platform. maybe arm/Kconfig is a better place.

qianfan-Zhao replied:
> XIP_EXECUTE_FROM_RAM
I prefer CONFIG_RAM_FUNCTION because the sections's name is ramfunc.
Also we can discussed more about this

qianfan-Zhao replied:
> Also, this is not, really, a Kernel option - rather a generic arch option, right? If you agree, I suggest to move the option to arch/Kconfig

Does the others SoC need ramfunc? if so I think should leave this to the kernel/Kconfig.

ioannisg replied:
> others

Allowing the firmware to execute code from SRAM has little or nothing to do with the Zephyr Kernel. That's why I suggested the option to be defined in arch/ instead.

ioannisg replied:
> I prefer CONFIG_RAM_FUNCTION because the sections's name is ramfunc.

What you are actually contributing here, is a way for the firmware to execute functions located in RAM. We need to propose a name that reflects exactly this thing.

ioannisg reviewed (changes requested):
I added some review comments and change requests.

ioannisg commented:
@qianfan-Zhao you could remove the WIP - I suppose we treat this work as any other PR to be reviewed...

qianfan-Zhao commented:
I had update the code based on the review:
```
diff --git a/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h b/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h
index 7967b5664a..294cc779d3 100644
--- a/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h
+++ b/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h
@@ -112,7 +112,8 @@ static inline void _get_mpu_ramfunc_region_attr(arm_mpu_region_attr_t *p_attr,
 {
 	ARG_UNUSED(base);
 
-	p_attr->rasr = _get_region_attr(0, RO, 0, 1, 0, 0, 0, size);
+	/* XXX */
+	p_attr->rasr = _get_region_attr(0, RO, 1, 1, 1, 0, 0, size);
 }
 #endif
 
diff --git a/arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h b/arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h
index f24c3e58bf..cc681fc13e 100644
--- a/arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h
+++ b/arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h
@@ -73,6 +73,16 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
 	p_attr->r_limit = REGION_LIMIT_ADDR(base, size);
 }
 
+#ifdef CONFIG_RAM_FUNCTION
+static inline void _get_mpu_ramfunc_region_attr(arm_mpu_region_attr_t *p_attr,
+						u32_t base, u32_t size)
+{
+	p_attr->rbar = ((ap << MPU_RBAR_AP_Pos) & MPU_RBAR_AP_Msk);
+	p_attr->mair_idx = MPU_MAIR_INDEX_SRAM;
+	p_attr->r_limit = REGION_LIMIT_ADDR(base, size);
+}
+#endif
+
 /**
  * This internal function is utilized by the MPU driver to combine a given
  * MPU RAM attribute configuration and region size and fill-in a structure with
diff --git a/include/arch/arm/cortex_m/scripts/linker.ld b/include/arch/arm/cortex_m/scripts/linker.ld
index b9b59f3594..c692daf4f4 100644
--- a/include/arch/arm/cortex_m/scripts/linker.ld
+++ b/include/arch/arm/cortex_m/scripts/linker.ld
@@ -263,6 +263,9 @@ SECTIONS
 
     GROUP_START(RAMABLE_REGION)
 
+#if defined(CONFIG_RAM_FUNCTION)
+#include <linker/ramfunc.ld>
+#endif
 
 #if defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_IS_BOOTLOADER)
     /* Must be first in ramable region */
@@ -414,10 +417,6 @@ SECTIONS
 
     __data_ram_end = .;
 
-#ifdef CONFIG_RAM_FUNCTION
-#include <linker/ramfunc.ld>
-#endif
-
     /* Define linker symbols */
 
     _image_ram_end = .;
diff --git a/include/linker/linker-defs.h b/include/linker/linker-defs.h
index 21a7945392..865fce7c7b 100644
--- a/include/linker/linker-defs.h
+++ b/include/linker/linker-defs.h
@@ -249,7 +249,12 @@ extern char _vector_end[];
 /* end address of image, used by newlib for the heap */
 extern char _end[];
 
-/* ramfunc sections */
+/* Memory owned by the kernel. Start and end will be aligned for memory
+ * management/protection hardware for the target architecture.
+ *
+ * All the functions with '__ramfunc' keyword will be placed into this
+ * sections.
+ */
 extern char __ramfunc_ram_start[];
 extern char __ramfunc_ram_end[];
 extern char __ramfunc_rom_start[];
diff --git a/include/linker/ramfunc.ld b/include/linker/ramfunc.ld
index c60f268c31..7b12e34dea 100644
--- a/include/linker/ramfunc.ld
+++ b/include/linker/ramfunc.ld
@@ -1,16 +1,19 @@
 SECTION_DATA_PROLOGUE(.ramfunc,,)
 {
-#ifdef CONFIG_CPU_HAS_MPU
+#ifdef CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT
 	. = ALIGN( 1 << LOG2CEIL(__ramfunc_ram_end - __ramfunc_ram_start));
 #endif
 	__ramfunc_ram_start = .;
 	KERNEL_INPUT_SECTION(.ramfunc)
 	KERNEL_INPUT_SECTION(".ramfunc.*")
-#ifdef CONFIG_CPU_HAS_MPU
+#ifdef CPU_HAS_ARM_MPU
 	. = ALIGN(32);
+#endif
+#ifdef CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT
 	. = ALIGN( 1 << LOG2CEIL(__ramfunc_ram_end - __ramfunc_ram_start));
 #endif
 	__ramfunc_ram_end = .;
+	. = ALIGN(32);
 } GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
 
 __ramfunc_rom_start = LOADADDR(.ramfunc);
```

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -73,6 +73,16 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
 }
 
+#ifdef CONFIG_RAM_FUNCTION
+static inline void _get_mpu_ramfunc_region_attr(arm_mpu_region_attr_t *p_attr,
```
Just to confirm with you, @qianfan-Zhao 
This attribute-setting shall be identical to _get_mpu_ram_region_attr(....), except for _not_ setting the XN bit, correct?

ioannisg replied:
BTW, this code will _not_ compile in; if there is no USERSPACE || MPU_STACK_GUARD || APPLICATION_MEMORY
Maybe you need to implement this function before line 59

qianfan-Zhao replied:
> This attribute-setting shall be identical to _get_mpu_ram_region_attr(....), except for not setting the XN bit, correct?

Yes

ioannisg replied:
This is still not good; we do not want all the rest of the implementation to be compiled in. See comment above.

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -263,6 +263,9 @@ SECTIONS
 
     GROUP_START(RAMABLE_REGION)
 
+#if defined(CONFIG_RAM_FUNCTION)
```
I think this is put _too_ early.
Would it be more correct to add it right before the CONFIG_APP_SHARED_MEM section?

SebastianBoe replied:
It does look like it's too early. Right below we have something saying

"Must be first in ramable region"

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -414,7 +417,6 @@ SECTIONS
 
     __data_ram_end = .;
 
-
```
This change (line removal) could be removed.

ioannisg commented on include/linker/linker-defs.h:
```diff
@@ -249,6 +249,16 @@ extern char _vector_end[];
+/* Memory owned by the kernel. Start and end will be aligned for memory
+ * management/protection hardware for the target architecture.
+ *
+ * All the functions with '__ramfunc' keyword will be placed into this
```
This text is nice.

ioannisg commented on include/linker/linker-defs.h:
```diff
@@ -249,6 +249,16 @@ extern char _vector_end[];
 /* end address of image, used by newlib for the heap */
 extern char _end[];
 
+/* Memory owned by the kernel. Start and end will be aligned for memory
```
Is this really owned by the kernel? @SebastianBoe ?

SebastianBoe replied:
Assuming he's talking about kernelspace, no, it is not in kernelspace.

qianfan-Zhao replied:
So in your opinion, who is the owner?

SebastianBoe replied:
I don't understand the question. Can you define owner?

It might be that we would want to say "Memory managed by Zephyr". Or just omit the statement, as most of the memory is managed by zephyr.

If you could try to explain what you want to say, then maybe I could help with how to say it :)

qianfan-Zhao replied:
The memory is doesn't owned by application, but if say it is managed by Zephyr, it's hard to understand, I think.

How about:
Memory are layout by the linker based on the MPU's requests.

SebastianBoe replied:
That sounds a bit better ...

But maybe:

Memory in the 'ramfunc' sections are placed by the linker based on the MPU's requirements.

qianfan-Zhao replied:
Thanks, changed.

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,19 @@
+{
+#ifdef CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT
+	. = ALIGN( 1 << LOG2CEIL(__ramfunc_ram_end - __ramfunc_ram_start));
+#endif
```
@SebastianBoe @andrewboie could you review the linker script?
Basically we need to satisfy
 - ARMv7M MPU with requirement for power-of-two alignment
- ARMv8-M with requirement for non-overlapping regions (aligned with 32-byte minimum)

ioannisg reviewed (changes requested):
Looks much better - still some issues to clarify at the linker script, and some minor issues/questions.

ioannisg commented on kernel/init.c:
```diff
@@ -167,6 +167,10 @@ void _data_copy(void)
 {
 	(void)memcpy(&__data_ram_start, &__data_rom_start,
 		 ((u32_t) &__data_ram_end - (u32_t) &__data_ram_start));
+#ifdef CONFIG_RAM_FUNCTION
```
A bit late on this, but i think we now have a convention of having only one underscore in the beginning of variables.
See _app_smem_start.
could you change this?

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -57,7 +57,7 @@ static void _region_init(u32_t index, struct arm_mpu_region *region_conf)
 
 #if defined(CONFIG_USERSPACE) || defined(CONFIG_MPU_STACK_GUARD) || \
-	defined(CONFIG_APPLICATION_MEMORY)
+	defined(CONFIG_APPLICATION_MEMORY) || defined(CONFIG_RAM_FUNCTION)
```
So, if CONFIG_RAM_FUNCTION is set, all the code below is compiled in, although we might only want _get_mpu_ramfunc_region_attr().

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,19 @@
+	. = ALIGN(32);
+} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
+
+__ramfunc_rom_start = LOADADDR(.ramfunc);
```
Something wrong with the new line at the end of the file?

qianfan-Zhao replied:
No, It's my mistake.

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,6 +282,10 @@ SECTIONS
 	}
 #endif
 
+#if defined(CONFIG_RAM_FUNCTION)
```
@SebastianBoe could you look at this, once more?
It looks like this section is, now, the very first section; so now _image_ram_start will not be the start of SRAM. Should this be OK, or requires a fix?

qianfan-Zhao commented:
@ioannisg I had updated, please review again.

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,20 @@
+SECTION_DATA_PROLOGUE(.ramfunc,,)
+{
+#ifdef CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT
+	. = ALIGN( 1 << LOG2CEIL(_ramfunc_ram_end - _ramfunc_ram_start));
```
@qianfan-Zhao below this line we need to align minimum with 32-byte, to avoid MPU issues with ARMv8-M, which requires non-overlapping regions, and 32-byte minimum MPU region

ioannisg reviewed (changes requested):
Looks, basically, good. Added a minor suggestion.
I would like @agross-linaro to have a look as well.

Note that, we should wait until #11528 gets merged; then we re-base this one and #11199

ioannisg commented:
@qianfan-Zhao could you please, rebase this one and make CI pass?

qianfan-Zhao commented:
recheck

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,19 @@
+/*
```
Does it make sense to have, now, an external linker file?
Can we do as for the NoCache PR, and overwrite the main ARM linker script?

qianfan-Zhao replied:
I don't know if the other platform need this, if so leave this as a external linker file is OK. If not, we'd better do that such as Nocache PR.

ioannisg replied:
I think the same applies to the no-cache section.
I believe it is OK to have this linker script code _inlined_ into the main arm linker script (in the very same place).

ioannisg replied:
Then, the linker script reader can actually see all sections in-place :)

aurel32 commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -95,6 +96,9 @@ static inline int _get_region_attr_by_type(arm_mpu_region_attr_t *p_attr,
+#ifdef CONFIG_RAM_FUNCTION
+	case RAM_FUNCTION_REGION:
+		_get_mpu_ramfunc_region_attr(p_attr, base, size);
 		return 0;
```
It looks like there is an issue there with a missing #endif. Probably a rebase issue.

qianfan-Zhao replied:
Thanks! Fixed.

aurel32 commented on include/arch/arm/cortex_m/mpu/arm_core_mpu_dev.h:
```diff
@@ -34,6 +34,8 @@ extern "C" {
 	NOCACHE_MEMORY_REGION,
+#ifdef CONFIG_RAM_FUNCTION
+	RAM_FUNCTION_REGION,
 #endif
```
It also looks like there is a missing #endif here.

qianfan-Zhao replied:
Fixed.

ioannisg commented on include/linker/ramfunc.ld:
```diff
@@ -0,0 +1,19 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+SECTION_DATA_PROLOGUE(.ramfunc,,)
```
Why not SECTION_PROLOGUE here?

ioannisg replied:
OK, it is for getting this data proplerly loaded, actually.

qianfan-Zhao commented:
recheck

andrewboie commented:
Hi, please rebase and re-work this patch.

aurel32 committed:
arch: arm: mpu: get the __ramfunc region size from the linker

The linker file defines the __ramfunc_ram_size symbols to get the size
of the __ramfunc_ram section. Use that instead of computing the value at
runtime from the start and end symbols. This saves 16 bytes of code with
CONFIG_RAM_FUNCTION=y.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
arch: make __ramfunc support transparent

Instead of having to enable ramfunc support manually, just make it
transparently available to users, keeping the MPU region disabled if not
used to not waste a MPU region. This however wastes 24 bytes of code
area when the MPU is disabled and 48 bytes when it is enabled, and
probably a dozen of CPU cycles during boot. I believe it is something
acceptable.

Note that when XIP is used, code is already in RAM, so the __ramfunc
keyword does nothing, but does not generate an error.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
I have been playing about `__ramfunc` (I need it to implement hwinfo driver on SAM E70), and I am not sure it is something that should depends on the MPU, at least on ARM:
- If the MPU is disabled, it is possible to execute code from RAM. Therefore we just need the linker and init code to support that
- If the MPU is enabled, we indeed need to configure the `__ramfunc` regions as executable

The new MPU infrastructure does not configure a region if it has a zero size with very low overhead (16 to 24 bytes of code wasted and a few cycles at boot time). I therefore think we can support `__ramfunc` on ARM independently on the MPU. I guess we can define a symbol like `ARCH_HAS_RAMFUNC_SUPPORT` that is always enabled on ARM and guard the code in `include/toolchain/gcc.h` with it. All the other code does not need to be guarded.

This has the advantage of providing `__ramfunc` support transparently to the users. Having less configuration options also means more chance to have that code tested.

ioannisg commented:
Apparently we need this for next release

SebastianBoe commented:
Would it be possible, when a function run from RAM is needed, to put it in a dedicated file, and use "CODE_DATA_RELOCATION" instead?

My concern is multiple things doing nearly the same thing with no re-use between them.

aurel32 commented:
> Would it be possible, when a function run from RAM is needed, to put it in a dedicated file, and use "CODE_DATA_RELOCATION" instead?

From what I get it will only allow to get rid of the changes in `include/toolchain/gcc.h` and `kernel/init.c`. I think there is still a need to provide a way to declare a function as `__ramfunc` in the headers to mark it with `__attribute__((long_call))`.

In addition to the technical aspect, `__ramfunc` is often used for a single critical function in a code unit, so it's not really practical to use that.

ioannisg commented:
@aurel32 @qianfan-Zhao would you, guys, be able to re-base this one, so we could finalize the review?

aurel32 commented:
@ioannisg I have rebased it locally sometimes ago in order to implement PR #13226. How should I submit that?

qianfan-Zhao commented:
@aurel32 I am busying another work and doesn't have enough time to continue this PR. It's greate if you can continue this PR. 😁

aurel32 commented:
> @aurel32 I am busying another work and doesn't have enough time to continue this PR. It's greate if you can continue this PR. grin

Thanks. I have just pushed your original commit rebased onto master and two additional commits to transparently enabled ramfunc support.

ioannisg commented:
Thanks for the rebasing @aurel32 , looks much clean now. We need some re-review here; @agross-linaro @galak could you also take a look?

Also, I wondering - should we have a kernel test case to cover this?

ioannisg commented on arch/Kconfig:
```diff
@@ -339,6 +339,9 @@ config ARCH_HAS_EXECUTABLE_PAGE_BIT
 config ARCH_HAS_NOCACHE_MEMORY_SUPPORT
 	bool
 
+config ARCH_HAS_RAMFUNC_SUPPORT
```
@andrewboie are you ok with this cross-arch Kconfig option?

andrewboie replied:
No not really, my issue is with the name.
Because this Kconfig is describing an architecture capability, and if the architecture is capable, then we automatically use up a static memory region for the ramfunc area even if there is nothing in it, this gets selected by another Kconfig.

Just call this CONFIG_RAMFUNC, and have it depend on CONFIG_XIP, and let the user turn it on if they want to.

Now it IS true that some XIP boards (like arduino 101) are incapable from fetching instructions from anywhere but flash for security reasons, but I don't think we really need a kconfig ontology for that.

aurel32 replied:
> Because this Kconfig is describing an architecture capability, and if the architecture is capable, then we automatically use up a static memory region for the ramfunc area even if there is nothing in it, this gets selected by another Kconfig.

That is not correct. If __ramfunc is not used, _ramfunc_ram_size is 0. In that case the MPU region is not configured (that's a nice side effect of the MPU rewrite), so no region is wasted.

andrewboie replied:
OK that works for me

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -77,8 +77,15 @@ void _arch_configure_static_mpu_regions(void)
+		{
+		.start = (u32_t)&_ramfunc_ram_start,
+		.size = (u32_t)&_ramfunc_ram_size,
+		.attr = K_MEM_PARTITION_P_RX_U_RX,
```
Not sure about un-privileged execution, here. This is a static region, perhaps we should align with the other static regions that are NA for nPRIV. @andrewboie what is your take?

(On the other hand, the entire flash on XIP images is open for execution for both priv/unpriv). But that's application and system code. In these RAMFUNC regions, do we intend to give them out "as-is" to user threads?

aurel32 replied:
For me the goal of `__ramfunc` is to be able to place and execute some code in RAM. It can be because the flash is temporarily unavailable (ie during flashing) or for performances reason. At least in the later case we probably want to be able to do that from user threads.

I therefore think that those regions should be aligned to the flash attributes like in this PR. That's why they are also not writable. Note that in practice they slightly differ on the cacheability (write through vs write back), but given the permissions, that doesn't make any difference.

ioannisg replied:
Alright

andrewboie replied:
We currently don't distinguish between kernel and user text areas, I'm OK with this policy.

ioannisg reviewed (commented):
Added a few comments to think of, most important, the priv-level of the region, and the need for developing a test for the feature

andrewboie commented on arch/arm/core/Kconfig:
```diff
@@ -23,6 +23,7 @@ config CPU_CORTEX_M
 	select ARCH_HAS_STACK_PROTECTION if ARM_MPU || CPU_CORTEX_M_HAS_SPLIM
 	select ARCH_HAS_USERSPACE if ARM_MPU
 	select ARCH_HAS_NOCACHE_MEMORY_SUPPORT if ARM_MPU && CPU_HAS_ARM_MPU && CPU_CORTEX_M7
+	select ARCH_HAS_RAMFUNC_SUPPORT
```
NACK, for reasons stated in other comment.
Just make this something that can be turned on in Kconfig, and not selected.

tyrantlizard77 reviewed (changes requested):
LGTM other than some issues I have with the Kconfig
